### PR TITLE
Refactor segmentation setup

### PR DESCRIPTION
**Pull request by Woazboat**
Changed segment setup code to be easier to read and moved all segmentation related definitions into "segment-definitions.h". Functionality is unchanged.

Moved common code for x86_64 thread register setup into a separate function to reduce code duplication. (Similar to how it is implemented in x86_32)

Removed reference to non-existent "tss_selector" in boot.32.c
Removed ArchThreadRegisters->ss0 on x86_64

### CODE REVIEW

File: arch/x86/32/common/source/ArchThreads.cpp [hunk 1]
```diff
@@ -23,9 +23,8 @@ void ArchThreads::setAddressSpace(Thread *thread, ArchMemory& arch_memory)
 
 void ArchThreads::createBaseThreadRegisters(ArchThreadRegisters *&info, void* start_function, void* stack)
 {
-  info = (ArchThreadRegisters*)new uint8[sizeof(ArchThreadRegisters)];
-  memset((void*)info, 0, sizeof(ArchThreadRegisters));
-  pointer root_of_kernel_paging_structure = VIRTUAL_TO_PHYSICAL_BOOT(((pointer)ArchMemory::getRootOfKernelPagingStructure()));
+  info = new ArchThreadRegisters();
```
Comment (dgruss) [inline arch/x86/32/common/source/ArchThreads.cpp]: is this guaranteed to null the members?

File: arch/x86/64/source/ArchCommon.cpp [hunk 2]
```diff
@@ -230,14 +231,14 @@ extern "C" void entry64()
 
   gdt_ptr.limit = sizeof(gdt) - 1;
   gdt_ptr.addr = (uint64)gdt;
-  asm("lgdt (%%rax)" : : "a"(&gdt_ptr));
+  asm("lgdt (%%rax)" : : "a"(&gdt_ptr)); // Load GDT
```
Comment (dgruss) [inline arch/x86/64/source/ArchCommon.cpp]: this comment should be removed

File: arch/x86/64/source/ArchCommon.cpp [hunk 3]
```diff
@@ -230,14 +231,14 @@ extern "C" void entry64()
...
   asm("mov %%ax, %%ds\n"
       "mov %%ax, %%es\n"
       "mov %%ax, %%ss\n"
       "mov %%ax, %%fs\n"
       "mov %%ax, %%gs\n"
       : : "a"(KERNEL_DS));
-  asm("ltr %%ax" : : "a"(KERNEL_TSS));
+  asm("ltr %%ax" : : "a"(KERNEL_TSS)); // Load TSS
```
Comment (dgruss) [inline arch/x86/64/source/ArchCommon.cpp]: this comment should be removed

File: arch/x86/64/source/ArchInterrupts.cpp [hunk 4]
```diff
@@ -175,7 +175,7 @@ extern "C" void arch_contextSwitch()
   }
   assert(currentThread->isStackCanaryOK() && "Kernel stack corruption detected.");
   ArchThreadRegisters info = *currentThreadRegisters; // optimization: local copy produces more efficient code in this case
-  g_tss.rsp0 = info.rsp0;
+  g_tss.rsp0 = info.rsp0; // Setup TSS to use kernel stack of new thread when a change to privilege level 0 occurs
```
Comment (dgruss) [inline arch/x86/64/source/ArchInterrupts.cpp]: this comment should be removed

File: arch/x86/64/source/boot.32.C [hunk 5]
```diff
@@ -141,17 +149,20 @@ extern "C" void entry()
...
   PRINT("Setup Segments...\n");
-  setSegmentDescriptor(1, 0, 0, 0, 0, 1, 0);
-  setSegmentDescriptor(2, 0, 0, 0, 0, 0, 0);
-  setSegmentDescriptor(3, 0, 0, 0, 3, 1, 0);
-  setSegmentDescriptor(4, 0, 0, 0, 3, 0, 0);
-  setSegmentDescriptor(5, -1U, (uint32) TRUNCATE(&g_tss) | 0x80000000, sizeof(TSS) - 1, 0, 0, 1);
+
+  SegmentDescriptor* gdt_p = (SegmentDescriptor*) TRUNCATE(&gdt);
+
+  setSegmentDescriptor(gdt_p + KERNEL_CS_INDEX, 0, 0, 0, D_T_CODE | D_DPL0 | D_PRESENT | D_G_PAGE | D_64BIT | D_READABLE);
```
Comment (dgruss) [inline arch/x86/64/source/boot.32.C]: looks inconsistent with the 32bit variant

Comment (dgruss) [reply arch/x86/64/source/boot.32.C]: not sure why the interface should be different than on 32bits

File: arch/x86/64/include/segment-definitions.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    uint8 zero          :2;
+    uint8 granularity   :1;
+    uint8 baseLH;
+    uint32 baseH;
+    uint32 reserved;
+}__attribute__((__packed__)) SystemSegmentDescriptor;
+
+typedef union
+{
+    DataSegmentDescriptor dataseg;
```
Comment (dgruss) [inline arch/x86/64/include/segment-definitions.h]: i can't find where they are used, are they unused?

File: arch/x86/32/common/source/ArchThreads.cpp (hunk 1 above)
Comment (Woazboat) [reply arch/x86/32/common/source/ArchThreads.cpp]: Yes. ```new ArchThreadRegisters()``` value-initializes members, which will set all built-in types to 0.
Just ```new ArchThreadRegisters``` without the parentheses wouldn't initialize the struct members.

File: arch/x86/64/include/segment-definitions.h (hunk 6 above)
Comment (Woazboat) [reply arch/x86/64/include/segment-definitions.h]: Base SWEB doesn't do much with segments aside from initializing them, so they aren't really used anywhere by default. It's more for future use and to replace the opaque `typeL`/`typeH`in the old definition.
setSegmentDescriptor() could use them, but it's not really necessary there.

**Comment by dgruss** [comment]
On 13.11.2017 21:42, Woazboat wrote:
> Base SWEB doesn't do much with segments aside from initializing them, so 
> they aren't really used anywhere by default. It's more for future use 

Then I would not merge this. There is no plan to add 
functionality/generic features to the baseline SWEB.

> and to replace the opaque |typeL|/|typeH|in the old definition.

I appreciate the idea behind this, but I think this is not the right 
direction. We should not add more code just in case someone might need 
it, but instead keep the baseline as simple and comprehensible as possible.

**Comment by Woazboat** [comment]
Hmm, I think having it clearly structured like this could help with understanding the code, but okay.
So, remove the (sub-)definitions again and change 32 bit to match the new interface? 
(I was going to do 32 bit afterwards once I knew if it was okay or not.)

**Comment by dgruss** [comment]
On 13.11.2017 21:57, Woazboat wrote:
> Hmm, I think having it clearly structured like this could help with 
> understanding the code, but okay.

Having superfluous code is never good. If someone wants additional 
information we have the lecture slides and the Intel documentation, no 
need to have it directly in the code if it is not used.

> So, remove the (sub-)definitions again and change 32 bit to match the 
> new interface?
> (I was going to do 32 bit afterwards once I knew if it was okay or not.)

Regarding the new interface, i don't see why passing gdt_p everytime is 
prettier than the previous variant where only the index was passed...

**Comment by fkirc** [comment]
To this end it might be also a good idea to remove huge pages support within ArchMemory.cpp. (currently huge pages seems to be used nowhere, but this is something which should be only be merged into devel and then into master once the winter term is finished)

**Comment by Woazboat** [comment]
> Regarding the new interface, i don't see why passing gdt_p everytime is prettier than the previous variant where only the index was passed...

Segment descriptors can occur in other places as well (LDT), so that would decouple segment descriptor creation from the GDT.

**Comment by dgruss** [comment]
On 13.11.2017 22:09, Woazboat wrote:
> Segment descriptors can occur in other places as well (LDT), so that 
> would decouple segment descriptor creation from the GDT.

We don't have an LDT and we are not planning to introduce one.

**Comment by dgruss** [comment]
On 13.11.2017 22:08, fkirc wrote:
> To this end it might be also a good idea to remove huge pages support 
> within ArchMemory.cpp.

Several years ago SWEB had a more complete support for huge pages.
I widely removed it for simplification.
The kernel still uses huge pages, I think some parts of it are therefore 
still required.
Parts that are not required should be removed, yes.

> but this is something which should be only be merged into devel and then 
> into master once the winter term is finished)

I think all of this is.
Nothing solves any problem of the students, it's just minor improvements.

**Comment by Woazboat** [comment]
If it's okay like it is now I'll close this PR and resubmit it to the devel branch

**Comment by dgruss** [comment]
there are still several changes that i would not approve, i like the part where numbers are replaced by defines, but i don't like unnecessary interface changes, removing unused stuff is good though.
for now i would close the pull request and i would suggest submitting the change with the defines into the devel branch.

**Pull request closed without merging**